tianshou/data/utils/converter.py

```diff
@@ -80,3 +82,64 @@ def to_torch_as(
     """
     assert isinstance(y, torch.Tensor)
     return to_torch(x, dtype=y.dtype, device=y.device)
+
+
+# Note: object is used as a proxy for objects that can be pickled
+Hdf5ConvertibleValues = Union[
+    int, float, Batch, np.ndarray, torch.Tensor, object, 'Hdf5ConvertibleType'
+]
+Hdf5ConvertibleType = Dict[str, Hdf5ConvertibleValues]
+
+
+def to_hdf5(x: Hdf5ConvertibleType, y: h5py.Group) -> None:
+    """Copy object into HDF5 group."""
+    for k, v in x.items():
+        # dicts and batches are both represented by groups
+        if isinstance(v, (Batch, dict)):
+            subgrp = y.create_group(k)
+            if isinstance(v, dict):
+                subgrp_data = v
+            else:
+                subgrp_data = v.__getstate__()
+                subgrp.attrs["__convert_to__"] = "Batch"
+            to_hdf5(subgrp_data, subgrp)
+        # numpy arrays and pytorch tensors are written to datasets
+        elif isinstance(v, np.ndarray):
+            y.create_dataset(k, data=v)
+            y[k].attrs["__data_type__"] = "numpy"
+        elif isinstance(v, torch.Tensor):
+            y.create_dataset(k, data=v.cpu().numpy())
+            y[k].attrs["__data_type__"] = "torch"
+        # ints and floats are stored as attributes of groups
+        elif isinstance(v, (int, float)):
+            y.attrs[k] = v
+        # resort to pickle for any other type of object
+        else:
+            int_data = np.fromstring(pickle.dumps(v), dtype="uint8")
+            y.create_dataset(k, data=int_data, dtype="uint8")
+            y[k].attrs["__data_type__"] = "pickle"
+
+
+def from_hdf5(
+    x: h5py.Group, device: Optional[str] = None
+) -> Hdf5ConvertibleType:
+    """Restore object from HDF5 group."""
+    # handle datasets
+    if isinstance(x, h5py.Dataset):
+        if x.attrs["__data_type__"] == "numpy":
+            y = np.empty(x.shape, dtype=x.dtype)
```

Comment:
I would suggest you to use the 'fully qualified type description'. It would be much easier to maintain and more clear to me. Here is an example:
```python
from pydoc import locate  # Standard library helper method
object_type = locate("torch.Tensor")  # or equally, "torch.tensor", the same goes for "numpy.array"/"numpy.ndarray" 
object = object_type(x, device=device)
```

Reply:
Sounds good. I agree, that the code can probably be simplified. However, I probably still need a separate clause for torch.Tensor as I have to copy to cpu first and convert to numpy.

Reply:
> However, I probably still need a separate clause for torch.Tensor as I have to copy to cpu first and convert to numpy.

Indeed. So, what you are doing right now is probably the way to go. Trying to generalize something that it not generic is often more harmful than the contrary. So maybe you should keep it like that.

Reply:
> In the case of a numpy array I think using read_direct could reduce the memory overhead. I'll have to test whether it really makes a difference though.

Do as you wish. To me speed and memory overhead/footprint is not a big deal in this case. Just keep it simple and clear, so that we can build on it more easily in the future.